### ISSUE TITLE
Border control: missing labels

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#42095]
### Description

In the Border control, some form elements are unlabelled. There's nothing that gives these form elements a name so they will be announced by screen readers only with their type. Users won't have a clue what these form elements are meant for and they will be forced to explore the surrounding UI to figure it out.

### Step-by-step reproduction instructions

- edit a block that has a padding control e.g.: Columns
- use the Chrome dev tools to inspect the source of the Border control that is displayed in the sidebar
- select the border input field (type=number)
- observe there's no associated label element or any aria-label / aria-labelledby
- in the Chrome dev tools, go to the Accessibility tab
- observe the accessible 'Name' under the computed properties is empty:
 
<img width="1332" alt="Screenshot 2022-07-01 at 16 02 44" src="https://user-images.githubusercontent.com/1682452/176914405-652744cc-3194-4365-a49c-e0c39a754b9d.png">

- repeat for the slider input (type=range) and observe it's unlabelled.
- additionally: the 'Border color and style picker' button doesn't have a tooltip so its accessible name is not visible for users (think at speech recognition software users)

- click the 'Unlink sides' button
- there are now 4 input fields, each one for a border side
- check their accessible name in the Chrome dev tools is empty
- check there's actually a visually hidden `<label>` element but it's not associated with the input field: there's no for'id attributes association so the input fields are actually unlabelled
- additionally: none of the 'Border color and style picker' buttons specifies whether it's for top, right, bottom, left  

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by mirka** [comment]
See also https://github.com/WordPress/gutenberg/pull/42118#issuecomment-1175921129

Here's the breakdown of tasks.

## BorderControl

#42348

- [x] Wrap the whole thing in a `fieldset` and change the label to a `legend`
- [x] Add a label to the `UnitControl`
- [x] Add a label to the `RangeControl`
- [x] Add a `ToolTip` to the "Border color and style picker" button
- [x] Update the `getByRole()` queries in the tests so they include a `name`

## BorderBoxControl

#42351

- [ ] Wrap the whole thing in a `fieldset` and change the label to a `legend`
- [ ] In unlinked mode, each `BorderControl` should be labelled (visually hidden), and also be wrapped in a `ToolTip` with the same label
- [ ] Update the `getByRole()` queries in the tests so they include a `name`

cc @ciampo @aaronrobertshaw

**Comment by aaronrobertshaw** [comment]
Thanks for the breakdown @mirka 👍 

I've created a couple of quick PRs hoping to address most of the tasks noted.

- https://github.com/WordPress/gutenberg/pull/42348 
    - Should tick most of the boxes quoted above
    - I wasn't too sure what the labelling should be in a couple of places here, so I've left a question on the PR itself
- https://github.com/WordPress/gutenberg/pull/42351
    - The component now has a `fieldset` as its wrapper and `legend` for its label
    - The split border side controls already get given labels e.g. `Top border` etc. Given the side border controls are within a `fieldset>legend` markup now, do they still need tooltips?
    - People using screen readers would already have the field identified via the `fieldset`, `legend`, and field label right?
    - Visually, the positioning of each side border control overlaying a border box would also indicate which border the control configures despite that control's label being hidden.
    
 I'm happy to make any further changes we need, I'm just a little unclear on what exactly is required.

**Comment by mirka** [comment]
>   * The split border side controls already get given labels e.g. `Top border` etc. Given the side border controls are within a `fieldset>legend` markup now, do they still need tooltips?
>   * People using screen readers would already have the field identified via the `fieldset`, `legend`, and field label right?
>   * Visually, the positioning of each side border control overlaying a border box would also indicate which border the control configures despite that control's label being hidden.

Good question. What I understood from https://github.com/WordPress/gutenberg/pull/42118#issuecomment-1175921129 is that, users of **speech recognition** software (not **screen reader**) need to know the actual name of the form element so they can use voice commands to navigate to it. So even if they can visually perceive that a control is for the top border, and even if it's accessibly labeled, they don't use a screen reader so they need a way to visually tell what the actual label is. Tooltips aren't ideal because they still require the user to focus on the field first, but at least if they do it once they'll have a chance to learn what the label is and can use voice commands from then on.

**Comment by aaronrobertshaw** [comment]
>  need to know the actual name of the form element so they can use voice commands to navigate to it

Thanks for the explanation. Updating the `BorderBoxControl` PR https://github.com/WordPress/gutenberg/pull/42351 with feedback from https://github.com/WordPress/gutenberg/pull/42348 is on my todo list today or tomorrow, so I'll take a run at adding the Tooltips as part of those efforts.

**Comment by afercia** [comment]
> they need a way to visually tell what the actual label is. Tooltips aren't ideal because they still require the user to focus on the field first, but at least if they do it once they'll have a chance to learn what the label is 

@mirka thanks, yes that's exactly the point. Tooltips aren't ideal but without them the only alternative for these users (who generally have some form of motor impairment) would be inspecting the source in the browser dev tools to discover the name of a control 🙁

**Comment by aaronrobertshaw** [comment]
A draft PR exploring [tooltips for inner `BorderControl` fields](https://github.com/WordPress/gutenberg/issues/42095#issuecomment-1183968034) within a `BorderBoxControl` can be found in: https://github.com/WordPress/gutenberg/pull/42661

**Comment by andrewhayward** [comment]
I've been digging around a bit with the border controls in the editor, and [as has been mentioned elsewhere](https://github.com/WordPress/gutenberg/pull/42661#issuecomment-1193736114), I'm not entirely convinced that the current set up provides "a good and intuitive UX". Apologies for the following, potentially somewhat off-topic, brain dump!

The initially presented view visually implies some sort of 'border' group, with an input labelled 'radius'...

<img src="/WordPress/gutenberg/assets/159848/6d0a54a3-0c04-486b-b7bb-1a39386f053e" alt="Screenshot of editor control panel, showing a hand-drawn 'border' group, containing an input associated by an arrow with a 'radius' label." width="490" />

This association breaks down though with the addition of the border size and style control (which itself is somewhat confusingly labelled "border" in the pop-up tools menu). At this point, is 'border' still a group legend, or a label for the new field?

<img src="/WordPress/gutenberg/assets/159848/35bb3f73-34ee-4df5-af0a-ad189eb5dd11" alt="Screenshot of editor control panel, showing an input associated by an arrow with a 'border' label, and another input associated by an arrow with a 'radius' label." width="490" />

This issue could potentially be resolved by adding a more explicit label to the size and style control, making it clear that "border" still applies to the group as a whole...

<img src="/WordPress/gutenberg/assets/159848/e23eb82e-c88c-4d50-8c03-d254dd8ae24c" alt="Screenshot of editor control panel, showing a hand-drawn 'border' group, containing input associated by arrows with 'size and style' and 'radius' labels." width="490" />

The whole notion falls apart a bit though, because in reality 'border' is not a group legend but a header, and 'radius' isn't a label but a legend. This issue becomes particularly clear when 'radius' (for example) is unlinked to show four separate inputs. At this point, 'radius' ceases to be a label for a single control, and instead acts as a group heading.

<img src="/WordPress/gutenberg/assets/159848/e63a6ded-f98f-48e9-bd1c-20e99c48a618" alt="Screenshot of editor control panel, showing a hand-drawn 'border' group, containing an input associated by an arrows with 'size and style' and a 'radius' group with four unlabelled inputs." width="490" />

Here it is particularly unclear what the separate 'radius' fields are labelled as, and even though it is somewhat discoverable via tooltip, 'top left' doesn't actually mean much when taken out of isolation.

**Comment by aaronrobertshaw** [comment]
Great breakdown of some of the issues around the border controls @andrewhayward 👍 

I wonder if some lessons or inspiration can be taken from the latest iteration of our spacing controls. They now have a dropdown menu beside the property to select the desired side, axis, or "custom" configuration. Could something like that allow better labelling or use of fieldset/legend?

<img width="281" alt="Screenshot 2023-09-01 at 11 45 00 am" src="https://github.com/WordPress/gutenberg/assets/60436221/9b3a5646-a603-4dca-a296-06459a8f7cd2">

After some very quick and dirty hacking in dev tools, it might look something like this:
<img width="280" alt="Screenshot 2023-09-01 at 11 51 03 am" src="https://github.com/WordPress/gutenberg/assets/60436221/f0d7b36b-eb8a-48e0-a536-2b061770db12">
<img width="280" alt="Screenshot 2023-09-01 at 12 08 45 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/21645f85-5d81-41d4-bd4a-adf9fec64664">
<img width="283" alt="Screenshot 2023-09-01 at 12 10 01 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/c1b4e8bd-48c8-4382-9c7f-b01b2c02a319">

Obviously, there are some issues with the resulting screenshots above but they convey the basic idea.

**Comment by afercia** [comment]
Personally, I'd think the root problem is a bit broader and it relates to the UI inconsistencies between mental models that, from an user perspective, are similar but in the UI are visually and semantically represented in different ways. 

For example, border and padding are very similar mental models, as in: _they both relate to the four sides of a box_. However, the editor UI for these settings is very differnt:

Padding: the expanded UI always shows the controls stacked horizontallu:

<img width="604" alt="padding" src="https://github.com/WordPress/gutenberg/assets/1682452/478c7b54-f6b8-4075-8cdf-814a4608fb29">

Bodrer: the 4 controls are placed in a way to visually represent the top, left, right, and bottom sides:

<img width="306" alt="border" src="https://github.com/WordPress/gutenberg/assets/1682452/3f76d300-ea0a-460a-b3cc-d2f41d3a6323">

Whether or not such a 'positional' UI is good for users (and I doubt it, given the cognitive load), it's inconsistent with the padding. From a user perspetive, the most intuitive concept is that both paddign and borders apply to the four sides of the box. Why the UI is different?

Radius: the mental model behind this setting is sligthly different, as radius applies to the corners of a box instead of to its sides. However, the UI shows the four controls in a grid with two rows and two columns:

<img width="299" alt="radius" src="https://github.com/WordPress/gutenberg/assets/1682452/f276f2ca-f9ad-42ee-a9b2-99fec24120fb">

Margin: similar to padding but it only applies to top ad/or bottom sides. It may display one control or two, stacked horizontally.

Overall, I think these 4 settings show too many options to users and they're inconsistent, adding too much cognitiv load for users. Idea;;y, they should be greatly simplified. I'm not sure implementing 'smart' solutions, like the boder 'positional' controls, is ideal. It may seem a smart solution at first but in the longer run it's just confusing. Personally, I'd vote for a gret simplification and for removing the alternative view. Providing just _one_ view, the simpler the better, and consistent design and UX for all these 4 settings would be likely a best option.

**Comment by afercia** [comment]
Additionally: labelling the input fields with icons isn't great from both an accessibility and usability perspective. It would be great to simply have visible label elements.

**Comment by ciampo** [comment]
Thank you @andrewhayward and @afercia for the feedback — I also agree that it would be good to have a more consistent and uniform set of controls dealing with the 4 sides of the box model. I'm going to cc @WordPress/gutenberg-design as this conversation may be particularly relevant.

**Comment by richtabor** [comment]
> Radius: the mental model behind this setting is sligthly different, as radius applies to the corners of a box instead of to its sides. However, the UI shows the four controls in a grid with two rows and two columns

It's the four "corners": top, right, bottom, left. 

> I wonder if some lessons or inspiration can be taken from the latest iteration of our spacing controls. They now have a dropdown menu beside the property to select the desired side, axis, or "custom" configuration. Could something like that allow better labelling or use of fieldset/legend?

I think leaning into this approach could be interesting. 

It's natural as we push forward in some areas to take those and apply them elsewhere. Worth exploring common patterns across these, as long as there's comprehension. We can have varying experiences for different controls, as long as they support the user. But I do think it's worth exploring on this front.

**Comment by jameskoster** [comment]
> I think leaning into this approach could be interesting.

Agreed. Another big benefit is that multiple controls can benefit from subsequent enhancements to the general pattern. 

One note: splitting padding/margin into horizontal/vertical by default serves common use cases that I'm not sure are shared by border. IE if we tried this, I'd consider the default control for border to manipulate all sides together, with options to select just one side, or custom for all sides. No linked top/bottom and left/right.

**Comment by youknowriad** [comment]
Closing as fixed: the BorderBoxControl's split controls now carry explicit labels ("Top border", "Left border", "Right border", "Bottom border") — verified in the component source. Please reopen if any border sub-control still lacks an accessible name.
